### PR TITLE
Deprecate `AdminInterface::toString()` and `AdminInterface::getObjectMetatadata()`

### LINKED ISSUE #5640: [RFC] Deprecate `AdminInterface::toString()` and their default implementation
## Feature Request
Deprecate `AdminInterface::toString()` and their default implementation.

https://github.com/sonata-project/SonataAdminBundle/blob/bc279a80e701c200251d802c7910ff782c2770e5/src/Admin/AdminInterface.php#L287-L292

https://github.com/sonata-project/SonataAdminBundle/blob/bc279a80e701c200251d802c7910ff782c2770e5/src/Admin/AbstractAdmin.php#L2444-L2455

This method should be replaced by defining `__toString()` in order to get the string representation for a given admin (maybe based on  `AdminInterface::getSubject()`).

Regarding the string representation for any object, which is the current responsibility of this method, the replacement should be provided by another service (I don't know which one).

### PR DESCRIPTION
**Pull request by phansys**
## Subject

Deprecate `AdminInterface::toString()` and `AdminInterface::getObjectMetatadata()`.

I am targeting this branch, because these changes are BC.

Closes #5640.

## Changelog

```markdown
### Added
- Added `AbstractAdmin::subjectAsString()`, `AbstractAdmin::getSubjectMetadata()`;
- Added `Sonata\AdminBundle\Util\AdminSubjectExtractor` class and "sonata.admin.subject_extractor" service.

### Deprecated
- Deprecated `AdminInterface::toString()`, `AdminInterface::subjectAsString()` should be used as replacement;
- Deprecated `AdminInterface::getObjectMetatadata()`, `AdminInterface::getSubjectMetadata()` should be used as replacement.
```

## To do

- [x] Add a Twig filter to get the string representation of an admin's subject;    
- [x] Update the tests;
- [x] Define how to update `AbstractAdmin::getObjectMetadata()`;

### CODE REVIEW

File: src/Admin/AdminInterface.php [hunk 1]
```diff
@@ -285,12 +285,21 @@ public function setRouteBuilder(RouteBuilderInterface $routeBuilder);
...
      * @return string
      */
     public function toString($object);
 
+    /**
+     * Returns the string representation for the admin's subject.
+     *
+     * @throws \LogicException if the admin's subject is not set
+     */
+    public function subjectToString(): string;
```
Comment (core23) [inline src/Admin/AdminInterface.php]: This is a BC break

File: src/Admin/AbstractAdmin.php [hunk 2]
```diff
@@ -2487,6 +2497,24 @@ public function toString($object)
         return sprintf('%s:%s', ClassUtils::getClass($object), spl_object_hash($object));
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function subjectToString(): string
```
Comment (core23) [inline src/Admin/AbstractAdmin.php]:
```suggestion
    public function getSubjectName(): string
```

Comment (phansys) [reply src/Admin/AbstractAdmin.php]: It sounds confusing to me, since the string representation of a subject can be represented by any property, not just for a name related concept.

Comment (core23) [reply src/Admin/AbstractAdmin.php]: subject**To**String sounds also confusing to me, because you don't provide any parameter. What about `subjectAsString`? :laughing:

File: src/Util/AdminSubjectExtractor.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Util;
```
Comment (greg0ire) [inline src/Util/AdminSubjectExtractor.php]: No `Util` namespace please, it's going to get cluttered with unrelated stuff, I guarantee it.

File: src/Admin/AdminInterface.php [hunk 4]
```diff
@@ -285,12 +285,23 @@ public function setRouteBuilder(RouteBuilderInterface $routeBuilder);
...
      * @return string
      */
     public function toString($object);
 
+    /**
+     * Returns the string representation for the admin's subject.
+     *
+     * @throws \LogicException if the admin's subject is not set
+     */
+    // public function subjectAsString(): string;
```
Comment (greg0ire) [inline src/Admin/AdminInterface.php]: Maybe introduce a new interface instead? That way the admin can implement it from day one. Or it could delegate that responsability to something else.

File: src/Resources/views/CRUD/base_edit.html.twig [hunk 5]
```diff
@@ -15,6 +15,8 @@ file that was distributed with this source code.
     {# NEXT_MAJOR: remove default filter #}
     {% if objectId|default(admin.id(object)) is not null %}
         {{ 'title_edit'|trans({'%name%': admin.toString(object)|truncate(15) }, 'SonataAdminBundle') }}
+        {# NEXT_MAJOR: Remove previous line and uncomment the following one #}
+        {# {{ 'title_edit'|trans({'%name%': admin.subjectAsString()|truncate(15) }, 'SonataAdminBundle') }} #}
```
Comment (greg0ire) [inline src/Resources/views/CRUD/base_edit.html.twig]: It would be great to have a twig extension just for this. That way, Twig does not know or care what object has the `subjectAsString()` method.

File: src/Twig/Extension/SonataAdminExtension.php [hunk 6]
```diff
@@ -92,16 +99,26 @@ public function __construct(
...
             );
         }
+
+        // NEXT_MAJOR: make AdminSubjectExtractor required
+        if (null === $subjectExtractor) {
+            @trigger_error(
+                sprintf('Argument 6 for %s() will be required with the 4.0 release.', __METHOD__),
+                E_USER_DEPRECATED
+            );
+        }
```
Comment (greg0ire) [inline src/Twig/Extension/SonataAdminExtension.php]: Or you could have a separate Twig extension just for this, that way there is no need for all the conditionals.

File: src/Resources/views/CRUD/base_edit.html.twig (hunk 5 above)
Comment (greg0ire) [reply src/Resources/views/CRUD/base_edit.html.twig]: Looks like that's already what you did above :thinking: why not here?

File: src/Resources/views/Block/block_search_result.html.twig [hunk 7]
```diff
@@ -49,8 +49,12 @@ file that was distributed with this source code.
                             {% set link = admin.getSearchResultLink(result) %}
                             {% if link %}
                                 <li><a href="{{ link }}">{{ admin.toString(result) }}</a></li>
+                                {# NEXT_MAJOR: Remove previous line and uncomment the following one #}
+                                {# <li><a href="{{ link }}">{{ result|subjectAsString(admin)|raw }}</a></li> #}
```
Comment (greg0ire) [inline src/Resources/views/Block/block_search_result.html.twig]: Ideally, `subjectAsString` should be able to work with just an admin code (so a string), so that we are not forced to have this responsability in the admin class. The Twig extension should be able to find the `AdminExtractor` implementation for that admin code without asking the admin object.

File: src/Admin/AdminInterface.php (hunk 4 above)
Comment (phansys) [reply src/Admin/AdminInterface.php]: I agree that these responsibilities should belong to another class.

File: src/Resources/views/Block/block_search_result.html.twig (hunk 7 above)
Comment (phansys) [reply src/Resources/views/Block/block_search_result.html.twig]: I decided to let the responsibility at the admin class in order to allow the user override the method to control the string representation for an admin's subject (keeping the current behavior).
I'm not against on remove this method from the admin class if we have another way to provide the same feature.

File: src/Resources/views/CRUD/base_edit.html.twig (hunk 5 above)
Comment (phansys) [reply src/Resources/views/CRUD/base_edit.html.twig]: IMO, the extension should be mainly focused on the admin context: an admin which have a subject.
I guess it could (optionally) work with standalone objects.

Regarding this question, I don't get what you mean:

>Looks like that's already what you did above thinking why not here?

File: src/Twig/Extension/SonataAdminExtension.php (hunk 6 above)
Comment (phansys) [reply src/Twig/Extension/SonataAdminExtension.php]: It would be fine.

File: src/Util/AdminSubjectExtractor.php (hunk 3 above)
Comment (phansys) [reply src/Util/AdminSubjectExtractor.php]: I completely agree. I initially choose this NS based on the simility with `ClassUtils::getClass()`. When we have matured the scope of these responsibilities, I think we could choose a better name.

File: src/Resources/views/CRUD/base_edit.html.twig (hunk 5 above)
Comment (greg0ire) [reply src/Resources/views/CRUD/base_edit.html.twig]: I meant that on the previous code snippet, you are using a new twig function instead of calling a method of the admin object

File: src/Resources/views/Block/block_search_result.html.twig (hunk 7 above)
Comment (greg0ire) [reply src/Resources/views/Block/block_search_result.html.twig]: I remember facing the same issue when extracting the breadcrumbs builder responsabilité, and I think there was something similar for the template registry system. Sorry for being vague, I'm on my phone. IMO a solution would be to have the same service being used for 90% of admins (one admin being identified by its admin code), and then the 10% that need "overrides" could use other implementations. So 1 admin has 1 admin code, many services that may or may not be shared with other admins, and hopefully, in the end, no God admin object.

File: src/Resources/views/CRUD/base_edit.html.twig (hunk 5 above)
Comment (phansys) [reply src/Resources/views/CRUD/base_edit.html.twig]: Oh, I see.
Because where I've used `subject_to_string` Twig filter, the admin has no subject.

File: src/Admin/AbstractAdmin.php [hunk 8]
```diff
@@ -2487,6 +2500,24 @@ public function toString($object)
...
+            throw new \LogicException(sprintf('Subject is not set for admin "%s".', $this->getCode()));
+        }
+
+        $subject = $this->getSubject();
+
+        if (\is_callable([$subject, '__toString']) && null !== $toString = $subject->__toString()) {
+            return $toString;
+        }
+
+        return sprintf('%s:%s', ClassUtils::getClass($subject), spl_object_hash($subject));
```
Comment (core23) [inline src/Admin/AbstractAdmin.php]: Can we use the primary key for the object here?

Comment (phansys) [reply src/Admin/AbstractAdmin.php]: I think that could be possible only for model managers that use and expose that feature.

**Comment by core23** [comment]
Any updated on this @phansys ?

**Comment by phansys** [comment]
> Any updated on this @phansys ?

Not yet, sorry about that. I'm away of the news and the work on the project these weeks. I'll try to have this up to date ASAP.

Thank you.
